### PR TITLE
added reptile example

### PR DESCRIPTION
**Pull request by ADMoreau**
This is an example of Model Agnostic Meta Learning using the reptile algorithm for the task of few shot learning using the omniglot dataset.

### CODE REVIEW

File: examples/vision/reptile.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+Title: Few-Shot learning with Reptile
+Author: [ADMoreau](https://github.com/ADMoreau)
+Date created: 2020/05/21
+Last modified: 2020/05/29
+Description: Reptile meta learning algorithm implemented in Keras
+
+"""
+
+"""
+The Reptile algorithm (https://arxiv.org/abs/1803.02999) was developed by OpenAI to
```
Comment (fchollet) [inline examples/vision/reptile.py:11]: Use markdown formatting for the links

File: examples/vision/reptile.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
+"""
+Title: Few-Shot learning with Reptile
+Author: [ADMoreau](https://github.com/ADMoreau)
+Date created: 2020/05/21
+Last modified: 2020/05/29
+Description: Reptile meta learning algorithm implemented in Keras
```
Comment (fchollet) [inline examples/vision/reptile.py:6]: meta-learning

Also, "implemented in Keras" is not necessary: this is keras.io :)

File: examples/vision/reptile.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+"""
+
+"""
+The Reptile algorithm (https://arxiv.org/abs/1803.02999) was developed by OpenAI to
+perform model agnostic meta-learning. Specifically, this algorithm was designed to
+quickly learn to perform new tasks with minimal training, a task known as few-shot
+learning. The algorithm works by taking metagradients from the training of mini-batches
+representing new tasks, and using these gradients for a meta training step.
+"""
```
Comment (fchollet) [inline examples/vision/reptile.py:16]: Please add a section about the dataset: what the data is, what the goal is, why it's interested to use Reptile on this dataset.

File: examples/vision/reptile.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+eval_iters = 5
+inner_iters = 4
+
+eval_interval = 1
+train_shots = 20
+shots = 5
+classes = 5
+
+"""
+## Dataset Functions
```
Comment (fchollet) [inline examples/vision/reptile.py:45]: Please add basic data visualization: the reader should immediately understand what you're training on, concretely. This is what gives meaning to the code example.

File: examples/vision/reptile.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+                "batch %d: train=%f test=%f" % (meta_iter, accuracies[0], accuracies[1])
+            )
+
+"""
+## Visualize Results
+"""
+
+# First, some preprocessing to smooth the training and testing arrays for display.
+window_len = 100
+w = eval("np.hamming(window_len)")
```
Comment (fchollet) [inline examples/vision/reptile.py:234]: Always use fully-spelled out variable names, e.g. width

File: examples/vision/reptile.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+urllib3.disable_warnings()
+train_dataset = Dataset(training=True)
+test_dataset = Dataset(training=False)
+
+"""
+## Build the Model
+"""
+
+
+class ModelLayer(layers.Layer):
```
Comment (fchollet) [inline examples/vision/reptile.py:140]: This layer is a simple combination of built-in layers, so prefer structuring it in a functional way:

```python
def conv_bn(x):
   x = layers.Conv2D(filters=64, kernel_size=3, strides=2, padding="same")(x)
   x = layers.BatchNormalization()(x)
   return layers.ReLU()(x)
```

File: examples/vision/reptile.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+        self.activation = layers.ReLU()
+
+    def call(self, inputs):
+        x = self.conv(inputs)
+        x = self.bnorm(x)
+        return self.activation(x)
+
+
+inputs = layers.Input(shape=(28, 28, 1))
+layer1 = ModelLayer()(inputs)
```
Comment (fchollet) [inline examples/vision/reptile.py:154]: Note that these are tensors, not layers, so they shouldn't be named "layer1". Use a different variable name

File: examples/vision/reptile.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+layer2 = ModelLayer()(layer1)
+layer3 = ModelLayer()(layer2)
+layer4 = ModelLayer()(layer3)
+flatten = layers.Flatten()(layer4)
+dense = layers.Dense(classes, activation="softmax")(flatten)
+# Log softmax
+outputs = tf.math.log(layers.Softmax()(dense))
```
Comment (fchollet) [inline examples/vision/reptile.py:161]: Taking the log of a softmax will lead to numerical instability. Either use softmax and train with `CategoricalCrossentropy()`, or do not apply any activation and train with `CategoricalCrossentropy(from_logits=True)`

File: examples/vision/reptile.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+Last modified: 2020/05/29
+Description: Reptile meta learning algorithm implemented in Keras
+
+"""
+
+"""
+The Reptile algorithm (https://arxiv.org/abs/1803.02999) was developed by OpenAI to
+perform model agnostic meta-learning. Specifically, this algorithm was designed to
+quickly learn to perform new tasks with minimal training, a task known as few-shot
+learning. The algorithm works by taking metagradients from the training of mini-batches
```
Comment (fchollet) [inline examples/vision/reptile.py:14]: Explain "meta-gradient" and "meta-training step"

File: examples/vision/reptile.py [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+            train_set, test_images, test_labels = dataset.get_mini_dataset(
+                eval_batch_size, eval_iters, shots, classes, split=True
+            )
+            old_vars = model.get_weights()
+            # Train on the samples and get the resulting accuracies.
+            for images, labels in train_set:
+                with tf.GradientTape() as tape:
+                    logits = model(images)
+                    loss = keras.losses.categorical_crossentropy(
+                        tf.one_hot(labels, depth=classes), logits
```
Comment (fchollet) [inline examples/vision/reptile.py:211]: Use SparseCategoricalCrossentropy so you don't have to one_hot.

File: examples/vision/reptile.py [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+        for dataset in (train_dataset, test_dataset):
+            # Sample a mini dataset from the full dataset.
+            train_set, test_images, test_labels = dataset.get_mini_dataset(
+                eval_batch_size, eval_iters, shots, classes, split=True
+            )
+            old_vars = model.get_weights()
+            # Train on the samples and get the resulting accuracies.
+            for images, labels in train_set:
+                with tf.GradientTape() as tape:
+                    logits = model(images)
```
Comment (fchollet) [inline examples/vision/reptile.py:209]: If these are supposed to be logits, then there should be no softmax and you should use `from_logits=True`

File: examples/vision/reptile.py [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+train_s = np.r_[
+    training[window_len - 1 : 0 : -1], training, training[-1:-window_len:-1]
+]
+test_s = np.r_[testing[window_len - 1 : 0 : -1], testing, testing[-1:-window_len:-1]]
+train_y = np.convolve(w / w.sum(), train_s, mode="valid")
+test_y = np.convolve(w / w.sum(), test_s, mode="valid")
+
+import matplotlib.pyplot as plt
+
+# And display the training accuracies.
```
Comment (fchollet) [inline examples/vision/reptile.py:244]: You are only display accuracy -- also display concrete examples. This is what gives the script its meaning.

File: examples/vision/reptile.py [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+
+class Dataset:
+    # This dataset function will fascilitate the creation of a few shot dataset
+    # from the Omniglot dataset that can be sampled from quickly while also
+    # allowing to create new labels at the same time.
+    def __init__(self, training):
+
+        # Download the tfrecord files containing the omniglot data and convert to a
+        # dataset.
+        if training:
```
Comment (fchollet) [inline examples/vision/reptile.py:57]: This class can generally be simplified quite a bit. One example would be these lines, where you only have to do:

```python
def __init__(self, split="train"):
   dataset = tfds.load("omniglot", split=split, as_supervised=True, shuffle_files=False)
```

Note that `batch(1)` is not useful since the dataset already yields samples one at a time.

File: examples/vision/reptile.py [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+        )
+        dataset = dataset.shuffle(100).batch(batch_size).repeat(repetitions)
+        if split:
+            return dataset, test_images, test_labels
+        return dataset
+
+
+import urllib3
+
+urllib3.disable_warnings()
```
Comment (fchollet) [inline examples/vision/reptile.py:131]: Why?

Comment (ADMoreau) [reply examples/vision/reptile.py:131]: Running it in Colab I get the following warnings 

/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py:847: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning)
/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py:847: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning)
/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py:847: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning)
/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py:847: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning)
/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py:847: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning)
/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py:847: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning)
/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py:847: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning)
/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py:847: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning)

I wasn't sure how to approach resolving these issues for someone who wants to quickly replicate the results.

Comment (fchollet) [reply examples/vision/reptile.py:131]: OK, please add a quick comment to explain that line.

Comment (ADMoreau) [reply examples/vision/reptile.py:131]: ok

File: examples/vision/reptile.py [hunk 15]
```diff
@@ -0,0 +1,295 @@
+"""
+Title: REPTILE
+Author: ADMoreau
```
Comment (fchollet) [inline examples/vision/reptile.py:3]: Do you want to add a personal link? (not mandatory)

**Comment by ADMoreau** [comment]
It's all ready!

**Comment by fchollet** [comment]
Please add the autogenerated files.

**Comment by ADMoreau** [comment]
Alright it should be ready to go.

**Comment by Himscipy** [comment]
Hello @ADMoreau ,
Thank you for this  Reptile examples (https://github.com/keras-team/keras-io/blob/master/examples/vision/reptile.py)
I have a question regarding few lines in the code.

can in place of line 197-202
```
for images, labels in mini_dataset:
        with tf.GradientTape() as tape:
            preds = model(images)
            loss = keras.losses.sparse_categorical_crossentropy(labels, preds)
        grads = tape.gradient(loss, model.trainable_weights)
        optimizer.apply_gradients(zip(grads, model.trainable_weights))
```

and  in line 220-226
```
# Train on the samples and get the resulting accuracies.
            for images, labels in train_set:
                with tf.GradientTape() as tape:
                    preds = model(images)
                    loss = keras.losses.sparse_categorical_crossentropy(labels, preds)
                grads = tape.gradient(loss, model.trainable_weights)
                optimizer.apply_gradients(zip(grads, model.trainable_weights))
```
model.fit be used ? 

In the current implementation, I am able to capture if line 220-226 is affecting the model improvement, since on line 230, the model weights are simply reseted.
```
# Reset the weights after getting the evaluation accuracies.
            model.set_weights(old_vars)
``` 
Any inputs will be really helpful. Thank you

**Comment by ADMoreau** [comment]
Yes, I believe you could replace the loops with a model.fit function so long as you only train for a single epoch.

**Comment by Himscipy** [comment]
Hi @ADMoreau , 
Thank you for response, it is really helpful. I wanted to check with you, if in the validation step (line 221-231), the data shown to the model is unique and has not been used in the training step. ? In addition, could you also clarify how the validation loop is been used to improvise. 
I am trying to adapt the implementation for a regression problem, I plan to use a completely new data/task not shown during the training at all to test if the trained weight would work well, i assume in line 265-272 you are doing that by updating the weights on the new task. If you could affirm that my understanding is correct it will be helpful. Thank you again.

**Comment by ADMoreau** [comment]
Happy to help. 

Validating a meta learning objective isn't the same as validation of a classification task. In meta learning you're showing the model's ability to learn a new task rapidly. Ergo, it does not matter if the data has not been seen before (in fact this is better proof of rapid learning capability) it only matters that the new data set is properly/ consistently labeled. Although this might get confusing because in this code I validate using a test-train split for a new sample of data for the validation loop, which is essentially what is happening in lines 265-272 as a final visualization demo of the effectiveness of the model.

If you only have two datasets where you are basically just testing your model's ability to extrapolate to a new distribution, meta learning is probably overkill. My understanding is that meta learning is used when you a large number of tasks and model needs to adapt rapidly to solve a new task several times. At no point in this code is the test data set a completely new dataset but, rather it is only a held out sample from the training dataset as described above used to validate a single iteration of meta learning. For more details of the training methodology I'd recommend the original paper https://arxiv.org/abs/1803.02999, particularly section 6.

**Comment by Himscipy** [comment]
Hello @ADMoreau,
Thank you for sharing the general insights and about the code.

If my last response, if I gave any impression that I have only two dataset, I would like to mention some corrections. In my case, the data is really big to train a model with it , but provides an opportunity to be posed as task specific training. The overall goal I have is to have a model which is generalizable enough to provide decent accuracy on unseen tasks. I have data for the unseen task and hence I plan to do the same as what have been done in line 265-272, where you retrain the model based on unseen data and then check the prediction. I completely  concur to your understanding "... is that meta learning is used when you a large number of tasks and model needs to adapt rapidly to solve a new task several times." 


As I mentioned in my earlier comment, I see that  you loop two dataset (train,test) as a loop, in the evaluation step inside the meta loop (line 221-231), is it mandatory to show the training data again ? or am I missing something here. Thank you

**Comment by ADMoreau** [comment]
That loop isn't necessary unless you want to evaluate your model, that is what the training dataset is doing here. Sorry to add more complexity here and that I likely misunderstood your questions earlier. 

The test dataset (as opposed to a test-train split from a subsample of the training dataset used in other steps) is a dataset held out to examine the model's ability to learn new distributions of data that are not in the main training loop. This is done because the main training loop dataset gets seen by the model far more frequently and so to avoid over fitting to a set of tasks, this test set of tasks is set aside.

The test-train tasks differentiation is just to examine the effectiveness of the model along the metrics I just outlined, you can do this anyway you wish or feel is an effective way to evaluate your model during training. Sorry if you already understood all of that but I want to make sure the logic is not ambiguous. It's my fault for not making better comments in the code.

**Comment by Himscipy** [comment]
Hi @ADMoreau , 

Thank you for your inputs and the explanation. 
No worries, about the code documentation, it is still clear. There are few things and my elementary experience in the area leading to query.

File: examples/vision/reptile.py [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+            old_vars = model.get_weights()
+            # Train on the samples and get the resulting accuracies.
+            for images, labels in train_set:
+                with tf.GradientTape() as tape:
+                    preds = model(images)
+                    loss = keras.losses.sparse_categorical_crossentropy(labels, preds)
+                grads = tape.gradient(loss, model.trainable_weights)
+                optimizer.apply_gradients(zip(grads, model.trainable_weights))
+            test_preds = model.predict(test_images)
+            test_preds = tf.argmax(test_preds).numpy()
```
Comment (ZhiliangWu) [inline examples/vision/reptile.py:228]: @ADMoreau Shall we add `axis=1` to `tf.argmax` here, so that we pick the largest index from different class predictions? Otherwise, we're picking the largest index from the batch direction.

Comment (ZhiliangWu) [inline examples/vision/reptile.py:228]: @ADMoreau @Himscipy `tf.argmax` by default would reduce the batch dimension (0). Shall we add `axis=1` to reduce the class dimension instead? As we want to pick the largest prediction from different class predictions.
